server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze;
+
+import io.crate.analyze.relations.AliasedAnalyzedRelation;
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.analyze.relations.AnalyzedRelationVisitor;
+import io.crate.analyze.relations.AnalyzedView;
+import io.crate.analyze.relations.DocTableRelation;
+import io.crate.analyze.relations.TableFunctionRelation;
+import io.crate.analyze.relations.UnionSelect;
+import io.crate.expression.symbol.AliasSymbol;
+import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.ScopedSymbol;
+import io.crate.expression.symbol.SymbolVisitor;
+import io.crate.metadata.Reference;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.Schemas;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.sql.tree.ColumnDefinition;
+import io.crate.sql.tree.ColumnType;
+import io.crate.sql.tree.Expression;
+import io.crate.sql.tree.TableElement;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/*
+ *  For the given Query(type: analyzedRelation), a list of tableElements will be returned according to its fields.
+ */
+public class ColumnDefinitionExtractor {
+
+    private final Schemas schemas;
+    //along the traversal, TableIdents or Aliases and its TableElements are stored for re-use
+    private final Map<RelationName, List<TableElement<Expression>>> TableIdentsToTableElementsMap = new HashMap<>();
+    private final QueryAnalyzedRelationVisitor queryAnalyzedRelationVisitor = new QueryAnalyzedRelationVisitor();
+    private final FieldsAstVisitor fieldsAstVisitor = new FieldsAstVisitor();
+
+
+    public ColumnDefinitionExtractor(Schemas schemas) {
+        this.schemas = schemas;
+    }
+
+    public List<TableElement<Expression>> extractFrom(AnalyzedRelation analyzedRelation){
+        return analyzedRelation.accept(queryAnalyzedRelationVisitor,null);
+    }
```

Comment:
I think I don't really understand why the visiting of the relations is necessary.

Wouldn't it be sufficient to do something like this:


```java
return Lists2.map(analyzedRelation.outputs(), this::outputToColumnDefinition)
```

And `outputToColumnDefinition` would look something like this:


```java
public static ColumnDefinition<Expression> outputToColumnDefinition(Symbol symbol) {
    ColumnIdent column = Symbols.pathFromSymbol(symbol);

    var type = MetadataToASTNodeResolver.dataTypeToColumnType(...);
    return new ColumnDefinition<>(
      column.name(),
      null,       -- default expression is always null
      null,       -- generated expression is always null
      type,
      List.of()   // constraints are always empty
    )
}
```

It looks like constraints are never copied from the source, so the above should be sufficient to handle all source relations in a generic way.



PostgreSQL:

```
psql=# create table tbl (x int default 10);
Time: 16.267 ms
psql=# create table copy as (select * from tbl);
Time: 3.355 ms
psql=# \d copy
                Table "public.copy"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 x      | integer |           |          |
```

Reply:
FWIW, the optional `like_option` clauses to the general `CREATE TABLE` statement [1] will allow users to specify which additional properties of the original table to copy.

The SO article at [2] outlines this a bit more and gives some examples.
```sql
-- Apply all properties from the source table.
CREATE TABLE IF NOT EXISTS <ClonedTableName> (LIKE <OriginalTableName> INCLUDING ALL);

-- Apply specific properties from the source table.
CREATE TABLE <ClonedTableName> (LIKE <OriginalTableName> INCLUDING DEFAULTS INCLUDING CONSTRAINTS INCLUDING INDEXES);
```

[1] https://www.postgresql.org/docs/current/sql-createtable.html
[2] https://stackoverflow.com/questions/11176383/create-table-as-with-primary-key-in-one-statement-postgresql

---

So, would the visiting of the relations as implemented by @jeeminso be suitable to reuse for implementing the `CREATE TABLE LIKE` feature instead?

Reply:
> So, would the visiting of the relations as implemented by @jeeminso be suitable to reuse for implementing the CREATE TABLE LIKE feature instead?

I think it requires a different implementation. `LIKE` is restricted to a table name, not an arbitrary relation, so it won't be necessary to handle different concrete `AnalyzedRelation` implementations

Reply:
> It looks like constraints are never copied from the source, so the above should be sufficient to handle all source relations in a generic way.

Yes, if constraints are not required, outputToColumnDefinition() should be sufficient.
One question, I tried it on MySQL and observed the constraints copied, does CrateDB references Postgres documents only?

```SQL
mysql> create table tbl (x int default 10);
Query OK, 0 rows affected (0.06 sec)

mysql> show create table tbl;
+-------+----------------------------------------------------------------------------------------+
| Table | Create Table                                                                           |
+-------+----------------------------------------------------------------------------------------+
| tbl   | CREATE TABLE `tbl` (
  `x` int(11) DEFAULT '10'
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+----------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> create table copy as (select * from tbl);
Query OK, 0 rows affected (0.06 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> show create table copy;
+-------+-----------------------------------------------------------------------------------------+
| Table | Create Table                                                                            |
+-------+-----------------------------------------------------------------------------------------+
| copy  | CREATE TABLE `copy` (
  `x` int(11) DEFAULT '10'
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+-----------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

Reply:
>  does CrateDB references Postgres documents only?

yes, we try to follow the SQL Standard and if there are behaviour extensions we try to mimic PostgreSQL.

Reply:
thanks for clarification, will drop the 'visiting relations' logic and proceed.

Reply:
> Also: I think for a first version it's okay to **not** support the LIKE clause. We can extend that later in separate PRs. Otherwise this one will become rather big.

Sure. I didn't want to argue otherwise and just wanted to support this conversation by giving a perspective into the other syntax which supports transferring table properties. Thanks for sorting this out, @jeeminso and @mfussenegger!

Reply:
I have spent some time on this and encountered one issue. If one of the outputs is a type of ScopedSymbol with its dataType being an ObjectType, its ColumnPolicy cannot be recovered without 'visiting relations'. Please help clarify whether recovering ColumnPolicy is required. 

I have tried to find Postgres equivalent of ObjectType and it seems that CompositeType is the closest but does not have ColumnPolicy feature.

Currently, I assigned ColumnPolicy.STRICT which should be default for ObjectTypes. 

Attached below the code snippet, in case you are interested. (almost all of the changes since the last changeset)
```java
    private List<TableElement<Expression>> getTableElementsFromSourceQuery(Query sourceQuery,
                                                                           ParamTypeHints paramTypeHints,
                                                                           CoordinatorTxnCtx txnCtx) {

        AnalyzedRelation analyzedSourceQuery = relationAnalyzer.analyze(
            sourceQuery,
            new StatementAnalysisContext(paramTypeHints, Operation.READ, txnCtx));

        return Lists2.map(analyzedSourceQuery.outputs(), CreateTableAsAnalyzer::outputToColumnDefinition);
    }

    private static ColumnDefinition<Expression> outputToColumnDefinition(Symbol symbol) {

        final ColumnIdent column = Symbols.pathFromSymbol(symbol);
        final ColumnPolicy columnPolicy = ColumnPolicy.STRICT;
        final DataType<?> dataType = symbol.valueType();
        var columnType = dataTypeToColumnType(dataType, columnPolicy);

        return new ColumnDefinition<>(
            column.name(),
            null,
            null,
            columnType,
            List.of()
        );
    }

    //The original version from MetadataToASTNodeResolver depends on its private Extractor 
    // that makes use of extractColumnDefinitions() and DocTableInfo
    private static ColumnType<Expression> dataTypeToColumnType(DataType<?> type,
                                                               ColumnPolicy columnPolicy) {

        if (type.id() == ObjectType.ID) {
            List<ColumnDefinition<Expression>> childColumns = new ArrayList<>();
            ((ObjectType) type).innerTypes().forEach(
                (childName, childDataType) -> {
                    childColumns.add(
                        new ColumnDefinition<>(
                            childName,
                            null,
                            null,
                            dataTypeToColumnType(childDataType, columnPolicy),
                            List.of())
                    );
                }
            );
            return new ObjectColumnType<>(columnPolicy.name(), childColumns);
        } else if (type.id() == ArrayType.ID) {
            DataType<?> innerType = ((ArrayType<?>) type).innerType();
            return new CollectionColumnType<>(dataTypeToColumnType(
                innerType,
                columnPolicy
            ));
        } else if (type.id() == StringType.ID) {
            StringType stringType = (StringType) type;
            if (stringType.unbound()) {
                return new ColumnType<>(type.getName());
            } else {
                return new ColumnType<>("VARCHAR", List.of(stringType.lengthLimit()));
            }
        } else {
            return new ColumnType<>(type.getName());
        }
    }

```

Reply:
I think for now it is okay to proceed with defaulting to `ColumnPolicy.STRICT`.

We may have to refactor the structures later to expose the column policy as part of the object-type, but we can follow up on that at a later point.